sky/provision/aws/utils.py

```diff
@@ -0,0 +1,148 @@
+"""Utils for AWS provisioner."""
+from typing import Dict, Any
+import threading
+
+import boto3
+from botocore import config
+import urllib3
+
+from ray.autoscaler._private.cli_logger import cf, cli_logger
+
+BOTO_MAX_RETRIES = 12
+
+# ======================== Thread-safe ========================
+# https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html#multithreading-or-multiprocessing-with-resources
+
+_lock = threading.RLock()
+
+
+def create_resource(resource: str,
+                    region: str,
+                    max_attempts: int = BOTO_MAX_RETRIES,
+                    **credentials) -> Any:
+    """Create an AWS resource."""
+    # overhead: 6.69 ms ± 41.5 µs
+    with _lock:
+        # We should prevent caching the resource for thread safety.
+        # Instead, the logic that uses the resource should reuse the same
+        # resource as much as possible.
```

Comment:
We do not need `threading.local()` for each thread, since we create a new resource under the lock each time, and it is only used by a single thread later. I measure the overhead of creating a new resource, it is not significant at all.